### PR TITLE
perf(fsdp_tp): reduce aten::copy_ overhead + add --loss-impl modes (chunked-backward, fused-linear, loss-parallel)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Reduces `aten::copy_` overhead in the FSDP+TP loss path and adds three selectable cross-entropy implementations via `--loss-impl`, each with an honestly-scoped operating range.

## New `--loss-impl` modes

- **`chunked-backward`** — custom autograd `Function` that bounds the tp=1 backward graph (saves ~one full logits buffer vs eager). Model-agnostic: works on any materialized logits incl. HF models, no `torch.compile` required. Good at moderate vocab/seq; does **not** fix the very-large-vocab two-buffer OOM.
- **`fused-linear`** (Liger / Cut-CE) — bounds both row and vocab dims via per-chunk checkpoint through the output module. Proven tp=1, ~32 GB. Needs the ezpz `Transformer` `return_hidden` path.
- **`loss-parallel`** — vocab-parallel CE with conditional `Shard(-1)` output. Proven tp=2. Comparison harness against `torch.distributed.tensor.parallel.loss_parallel`.

## Also

- `return_hidden` path on `Transformer.forward` (`models/llama.py`)
- Single `masked_fill` for the label ignore-mask (drops a redundant copy)
- Equivalence test harness across all loss-impl modes (chunk-size invariance, all-ignored rows)
- Docs scoping each mode's correct regime

## Testing

11 passed / 7 env-skipped (`tests/test_loss_impls.py`).

## Stats

11 commits · 4 files · +808 / −51

## Summary by Sourcery

Introduce multiple memory-optimized cross-entropy implementations for FSDP+TP training and wire them into the training loop via a new --loss-impl selector.

New Features:
- Add chunked-backward cross-entropy mode that bounds the backward graph while remaining model-agnostic and torch.compile-free.
- Add fused-linear cross-entropy mode that computes loss from per-chunk hidden states without materializing full logits, leveraging the Transformer return_hidden path.
- Add loss-parallel cross-entropy mode that performs vocab-parallel CE across tensor-parallel ranks using local shards and TP collectives.
- Extend the Transformer forward API with a return_hidden option to expose normalized hidden states for advanced loss implementations.

Bug Fixes:
- Reduce redundant aten::copy_ usage in label masking by consolidating multiple clones and assignments into a single masked_fill on an ignore mask.

Enhancements:
- Expand --loss-impl CLI help and behavior to describe and gate each loss mode appropriately based on model type and tensor-parallel configuration.
- Update tensor-parallel parallelize logic to optionally keep logits vocab-sharded and locally returned when using loss-parallel.
- Adjust training loop loss computation to select eager, chunked, chunked-backward, compiled, fused-linear, or vocab-parallel implementations based on configuration while preserving correctness.
- Skip logits debug probing when using fused-linear since the model returns hidden states and never materializes full logits.

Documentation:
- Document all cross-entropy implementation modes, their regimes, and CLI usage in the FSDP+TP example guide, including updated usage snippets.

Tests:
- Add a comprehensive equivalence test harness validating loss and gradient parity between eager and chunked, chunked-backward, compiled, fused-linear, and vocab-parallel implementations, including chunk-size invariance and all-ignored-row behavior.

### CODE REVIEW

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -511,6 +833,10 @@ def _compute_loss(
         return _cross_entropy_chunked(
             logits, labels, ignore_index=ignore_index, chunk_size=chunk_size
         )
+    if impl == "chunked-backward":
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:847]: Fixed in 0d6f51c. When a specialized path is disabled, `train()` now normalizes `args.loss_impl = "compiled"` so the call site runs the advertised compiled fallback instead of an unhandled impl. As defense-in-depth, `_compute_loss` now raises `ValueError` on an unknown impl rather than silently running eager CE (verified: `fused-linear`/`loss-parallel`/unknown all raise).

File: src/ezpz/examples/fsdp_tp.py [hunk 2]
```diff
@@ -1870,6 +2243,26 @@ def train(
...
+    if want_fused_linear:
+        has_output = hasattr(base_model, "output") and hasattr(
+            base_model.output, "weight"
+        )
+        if is_hf_model or not has_output:
+            logger.warning(
+                "--loss-impl=fused-linear needs the ezpz Transformer "
+                "(hidden states + output weight); falling back to compiled "
+                "for this model."
+            )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2301]: Good catch — fixed in 0d6f51c. The fused-linear fallback branches (HF/no-output, and tp>1) now set `args.loss_impl = "compiled"`, so `_compute_loss` actually runs compiled CE. Previously it left `loss_impl="fused-linear"`, which fell through to eager and reintroduced the OOM. `_compute_loss` also now raises on unknown impl so this can't regress silently.

File: src/ezpz/examples/fsdp_tp.py [hunk 3]
```diff
@@ -1870,6 +2243,26 @@ def train(
...
+        elif args.tp > 1:
+            logger.warning(
+                "--loss-impl=fused-linear with tp>1 is not yet supported "
+                "(needs vocab-shard composition); falling back to compiled."
+            )
+        else:
+            use_fused_linear = True
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2313]: Fixed in 0d6f51c — same root cause as the other fallback comments. Disabled fused-linear now sets `args.loss_impl = "compiled"` (real compiled fallback), and `_compute_loss` raises on unknown impl instead of defaulting to eager.

File: src/ezpz/examples/fsdp_tp.py [hunk 4]
```diff
@@ -1535,6 +1894,16 @@ def train(
     world_size = ezpz.distributed.get_world_size()
     assert world_size % args.tp == 0, "WORLD_SIZE must be divisible by TP"
     dpsize = world_size // args.tp
+    # loss-parallel (vocab-sharded CE) only does anything at tp>1; at tp=1 the
+    # local vocab IS the full vocab, so the loss call site falls back to eager.
+    use_loss_parallel = (
+        getattr(args, "loss_impl", "eager") == "loss-parallel" and args.tp > 1
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Fixed in 0d6f51c. `use_loss_parallel` is now resolved *after* the HF branch forces `args.tp = 1` (it was computed before). So an HF model launched with `--tp>1 --loss-impl=loss-parallel` no longer enters vocab-parallel CE with a stale TP group — tp is 1 by then, so it falls back and `args.loss_impl` is normalized to `compiled`.

File: src/ezpz/examples/fsdp_tp.py [hunk 5]
```diff
@@ -479,6 +488,319 @@ def _cross_entropy_chunked(
...
+        chunk = (global_vocab_size + tp_world - 1) // tp_world
+        vocab_start = min(global_vocab_size, chunk * tp_rank)
+        vocab_end = min(global_vocab_size, vocab_start + chunk)
+        local_vocab = max(0, vocab_end - vocab_start)
+        if logits_2d.shape[-1] != local_vocab:
+            raise ValueError(
+                "_VocabParallelCrossEntropy expected local vocab "
+                f"{local_vocab} for global {global_vocab_size}, got "
+                f"{logits_2d.shape[-1]}."
+            )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:666]: Fixed in 0d6f51c. `_VocabParallelCrossEntropy` now fails fast with a clear message when `tp_world > global_vocab_size` (which would otherwise produce an empty `local_vocab==0` shard and blow up later in gather/indexing).

File: tests/test_loss_impls.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+def test_vocab_parallel_matches_eager_2rank():
+    import torch
+    import torch.multiprocessing as mp
+
+    try:
+        mgr = mp.Manager()
+        ret = mgr.dict()
+        mp.spawn(
+            _vp_worker, args=(2, _VP_V, _VP_N, ret), nprocs=2, join=True
+        )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_loss_impls.py:280]: Fixed in 0d6f51c. The parent now picks a free port (bind to `:0`, read it back) and passes it into the worker, which sets `MASTER_PORT` explicitly instead of `setdefault`-ing a hard-coded value. No more inherited-busy-port hangs.

File: tests/test_loss_impls.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+@pytest.mark.skipif(not LOSS_AVAILABLE, reason="ezpz.examples.fsdp_tp not importable")
+class TestLossImplEquivalence:
+    @pytest.mark.parametrize("impl", LOGIT_IMPLS)
+    @pytest.mark.parametrize(
+        "shape", [(2, 16, 50), (1, 257, 128), (3, 8, 1000)], ids=["small", "long", "wide"]
+    )
+    def test_matches_eager_loss_and_grad(self, impl, shape):
+        if impl == "compiled":
+            pytest.importorskip("torch._dynamo")
+        torch.manual_seed(0)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_loss_impls.py:52]: Done in 0d6f51c. The vocab-parallel test is now parametrized over `vocab=[12, 13]` — 13 with world_size=2 gives a 7+6 split, exercising the uneven-shard math (`local_vocab` computation, out-of-range label clamping, ragged gather). Both cases pass against the eager full-vocab loss + grad-shard reference (16 passed / 3 skipped locally).

File: src/ezpz/examples/fsdp_tp.py [hunk 8]
```diff
@@ -1535,6 +1894,16 @@ def train(
     world_size = ezpz.distributed.get_world_size()
     assert world_size % args.tp == 0, "WORLD_SIZE must be divisible by TP"
     dpsize = world_size // args.tp
+    # loss-parallel (vocab-sharded CE) only does anything at tp>1; at tp=1 the
+    # local vocab IS the full vocab, so the loss call site falls back to eager.
+    use_loss_parallel = (
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Thanks — sensible suggestions, but deferring both for this PR to keep the diff focused on the correctness fixes (the fallback bug + hardening landed in 0d6f51c). The `LossStrategy` centralization and the softmax-minus-one-hot grad rewrite are pure refactors with no behavior change; tracking them as a follow-up cleanup rather than expanding this PR's review surface.

File: src/ezpz/examples/fsdp_tp.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Addressed. `use_loss_parallel` is now resolved *after* the HF branch forces `args.tp = 1`, and the stale `loss-parallel` impl is normalized to `compiled` at that point (`src/ezpz/examples/fsdp_tp.py:2128-2144`). The HF + `--tp>1 --loss-impl=loss-parallel` case is explicitly covered by this override, so it no longer enters vocab-parallel CE with a stale TP group. Resolving.

File: src/ezpz/examples/fsdp_tp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2301]: Addressed. Both fused-linear fallback branches now set `args.loss_impl = "compiled"` (`src/ezpz/examples/fsdp_tp.py:2305` for HF/no-output, `:2311` for tp>1), so the `_compute_loss` call site actually runs compiled CE instead of falling through. As a belt-and-suspenders, `_compute_loss` now raises `ValueError` on any unhandled impl (`:865`) rather than silently running eager CE + reintroducing the full-logits OOM. Resolving.

File: src/ezpz/examples/fsdp_tp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:2313]: Fixed. When `use_fused_linear` stays False the fallback branches now reassign `args.loss_impl = "compiled"` (`src/ezpz/examples/fsdp_tp.py:2305`, `:2311`), so the loss call runs compiled CE — not eager. And `_compute_loss` no longer treats unknown impls as eager: it raises `ValueError` (`:865`), so a missed normalization fails loudly instead of silently OOMing. Resolving.

File: src/ezpz/examples/fsdp_tp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py:666]: Fixed. `_VocabParallelCrossEntropy.forward` now fails fast when `tp_world > global_vocab_size` with a clear message pointing at `--tp` / `--loss-impl` (`src/ezpz/examples/fsdp_tp.py:650-656`), before any gather/indexing can hit an empty shard. Resolving.

File: tests/test_loss_impls.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_loss_impls.py:280]: Fixed. The test now binds a free port in the parent via a `_free_port()` helper (`tests/test_loss_impls.py:208`) and passes it into the workers before `mp.spawn` (`:274`), rather than `setdefault`-ing a hard-coded value that could inherit a busy `MASTER_PORT` from the environment. Resolving.
